### DISCUSSION TITLE
before and after data augmentation: shape changes plus the number of batches

### DISCUSSION THREAD

**Discussion by Joeycho** [MIC-DKFZ/nnUNet#1175] [Q&A] [answered]
Dear all,

Training and inference with nnUNet were very successful (Dental 3D medical data which is around teeth and jaws, (T1, T2, CT)), but that guided me to dig in what actually happened in the nnUNet pipeline.

So, I have only 11 3D training images (T1) and sometimes I put CT as well. T2 was not useful when I see the result T2 only and T1 and T2.

preprocessed data size is ([701, 360, 683])

Data loader do also crop the patch out of images enforcing half of batches to have class label for overcoming class imbalance. This patch size is ([219, 191, 257]).

Before the training(precisely network), data should be prepared as patches, in my case ([160  96 160]) as patch size. This is done by several data augmentation transforms and via `get_moreDA_augmentation` function.

Briefly the process for only images like the below, if I understood correctly,

([batch_size, modalities, 701, 360, 683])  1)->    ([batch_size, modalities, 219, 191, 257])    2)->   ([batch_size, modalities, 160  96 160])

My first question is in second arrow step 2), data augmentation is actually 'one to one' transformation, right? Of course, inverse transformation might not be possible in some cases, but I am asking..  **one** batch ([batch_size, modalities, 219, 191, 257]) is transformed to **one** after-data augmented batch ([batch_size, modalities, 160  96 160]). So, Data augmentation in nnUNet does not increase the number of data(samples). It is just **one toward the other** form, transformation and **increases the variability** in generated patches.

My second question is in first 1) arrow step, which happened in DataLoader3D. I enumerate DataLoader3D to check how many items will be in for 1 epoch. But I observed that it did not stop (keep increasing over 10000 items..) and I checked [another issue](https://github.com/MIC-DKFZ/nnUNet/issues/61#issuecomment-530267717) regarding **the number of batches**  So, I understood it as..  1) arrow step can **generate the unlimited number of batches** while preserving visibility of classes. And as default value, in training time, the number of batches is just fixed as 250 and in validation, it is fixed as 50. So as Fabian explained in [the issue](https://github.com/MIC-DKFZ/nnUNet/issues/61#issuecomment-530267717), 1 epoch in patch-based training is not the same as typical epoch definition. it does not guarantee covering all possible patches from training data. There could be replacement and duplicate ones, and it is heuristically defined, like 250 or 50..the number of batches in 1 epoch. **Does this mean, are the sets of generated patches in every epoch in nnUNet different each other?** For example, 250 batches from 1st epoch, and another 250 batches from 2nd epoch... and so on. If this is true, it seems that nnUNet prevents(or reduces) overfitting possibilities on training sets in its pipeline. Of course, we may be able to create trivia overfitting examples, tho.

In short, 1st question was whether data augmentation in nnUNet increase the number of patches(data) or not in training time.

2nd question was whether the number of batches is totally heuristic or not. And the meaning of that with the context of Epoch and overfitting.

**Comment by FabianIsensee** [comment, marked as answer]
> Does this mean, are the sets of generated patches in every epoch in nnUNet different each other?

each patch is sampled randomly and augmented randomly. I can confidently say that no patch will ever be the same as any other patch nnU-Net will generate in 1000 years. Too much of a combinatoric explosion. 

> So, Data augmentation in nnUNet does not increase the number of data(samples). It is just one toward the other form, transformation and increases the variability in generated patches.

The dataloading pipeline samples random patches and puts them in a batch. The batch is then augmented 1:1, no additional samples are generated.

Epoch in nnU-Net does not have a meaning. It's just an arbitrary number of iterations over batches.

Best,
Fabian

**Reply by Joeycho** [reply]
Super, thanks for understanding my questions and the perfect clarification :)  I'm working on creating simplifed UNet with all the results(hyperparameters) from nnUNet, but I keep finding more and more.. special things from nnUNet framework, which makes me quite challenging.. The network architecture itself is really simple (Generic UNet -> actual exact UNet which is used), but many interesting parts in preprocessing. Plus, nnUNet is surprisingly fast and use GPU RAM efficiently. I think in naive approach (loading images, no saving.. unpacking... as nnUNet), I cannot achieve that speed. I did it with DynUnet from MONAI, which is inspired by nnUNet, but it took way longer... (90 mins for 1 epoch(4 iterations...)). I just wanted to highlight that another big contribution of nnUNet is the speed of training and efficient usage of GPUs.

**Reply by FabianIsensee** [reply]
Thanks :-) Well we spent a lot of time optimizing things and many of the design choices in nnU-Net are not consistent with other frameworks. So it's not intuitive for a lot of people but the performance is definitely there
